pkg/urbit/jets/e/en_json.c

```diff
@@ -0,0 +1,438 @@
+    
+    while ((buf_u->len_w + len_w) >= buf_u->siz_w) {
+      buf_u = _resize_buffer(buf_u);
+    }
+
+    memcpy(&(buf_u->buf_y[buf_u->len_w]), buf_y, len_w);
+    buf_u->len_w += len_w;
+    return buf_u;
+  }
+
+  json_buffer*
+  _serialize_null(json_buffer *buf_u)
+  {
+    _append_text(buf_u, _JSON_NULL, 4);
+  }
+
+  json_buffer*
+  _serialize_true(json_buffer *buf_u)
+  {
+    _append_text(buf_u, _JSON_TRUE, 4);
+  }
+
+  json_buffer*
+  _serialize_false(json_buffer *buf_u)
+  {
+    _append_text(buf_u, _JSON_FALSE, 5);
+  }
+
+  json_buffer*
+  _serialize_loobean(json_buffer *buf_u, u3_noun a)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+
+    if ( a == c3y ) {
+      return _serialize_true(buf_u);
+    }
+    else if ( a == c3n ) {
+      return _serialize_false(buf_u);
+    }
+    else {
+      return _encode_fail(buf_u);
+    }
+  }
+
+  json_buffer*
+  _serialize_number(json_buffer *buf_u, u3_noun a)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+    if ( _(u3du(a)) ) {
+      return _encode_fail(buf_u);
+    }
+
+    const c3_y *byt_y;
+    // XX assumes little-endian
+    //
+    if ( c3y == u3a_is_cat(a) ) {
+      byt_y = (c3_y*)&a;
+    }
+    else {
+      u3a_atom* vat_u = u3a_to_ptr(a);
+      byt_y = (c3_y*)vat_u->buf_w;
+    }
+
+    buf_u = _append_text(buf_u, byt_y, u3r_met(3, a));
+    return buf_u;
+  }
+
+  json_buffer*
+  _serialize_string(json_buffer *buf_u, u3_noun a)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+    if ( _(u3du(a)) ) {
+      return _encode_fail(buf_u);
+    }
+
+    c3_w len_w = u3r_met(3, a);
+
+    buf_u = _append_char(buf_u, '"');
+    for (c3_w i = 0; i < len_w; ++i) {
+      c3_y c_y = u3r_byte(i, a);
+
+      switch ( c_y ) {
+        case 0 ... 9:
+        case 11 ... 31:
+        {
+          buf_u = _append_text(buf_u, _JSON_UNICODES[c_y], 6);
+          break;
+        }
+
+        case 10:
+        {
+          buf_u = _append_text(buf_u, _JSON_NEWLINE, 2);
+          break;
+        }
+
+        case 34:
+        {
+          buf_u = _append_text(buf_u, _JSON_DOQ, 2);
+          break;
+        }
+
+        case 92:
+        {
+          buf_u = _append_text(buf_u, _JSON_BAS, 2);
+          break;
+        }
+
+        case 127:
+        {
+          buf_u = _append_text(buf_u, _JSON_DEL, 6);
+          break;
+        }
+
+        default:
+        {
+          buf_u = _append_char(buf_u, c_y);
+        }
+      }
+    }
+    buf_u = _append_char(buf_u, '"');
+
+    return buf_u;
+  }
+
+  json_buffer*
+  _serialize_array(json_buffer *buf_u, u3_noun a)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+
+    buf_u = _append_char(buf_u, '[');
+
+    if ( u3_nul != a ) {
+      u3_noun i, t = a;
+
+      while ( u3_nul != t ) {
+        u3x_cell(t, &i, &t);
+        buf_u = _serialize(buf_u, i);
+        buf_u = _append_char(buf_u, ',');
+      }
+
+      if ( buf_u == NULL )
+        return NULL;
+      
+      --buf_u->len_w;
```

Comment:
Can you leave a comment noting that this handles the trailing comma?

Reply:
Done